src/Compilers/Core/Portable/Operations/IDataFlowAnalyzer.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis
+{
+#if WORKSPACE
+    using TBasicBlock = BasicBlock;
+#else
+    using TBasicBlock = BasicBlockBuilder;
+#endif
+
+    /// <summary>
+    /// Analyzer to execute custom dataflow analysis on a control flow graph.
+    /// </summary>
+    /// <typeparam name="TBlockAnalysisData">Custom data tracked for each basic block with values at start of the block.</typeparam>
+    internal interface IDataFlowAnalyzer<TBlockAnalysisData> : IDisposable
```

Comment:
`IDataFlowAnalyzer` type to implement custom dataflow analysis.

Reply:
is this an internal file that take compiler logic and tries to embed it directly into the workspace?  I cannot imagine how this is a suitable approach for doing this work.  This seems to be the antithesis of how Roslyn is intended to be used.  It keeps the compiler a black box system, but just privileges the IDE to be able to leverage literal raw code within it.  This is similar to the old native approach we used to have and which we moved away from.  

If, however, this PR is simply to prove out teh API you want ot make public, then this is ok *only* as something *not* to be checked in.  Effectively, the PR would be: here's the API shape we think is right, and here's a feature that shows that it works.

Then, once the API is approved and made public, we can then have the IDE PR that then sits on that public API.

So, in other words, this is ok as a transitory *feature branch* feature.  But not ok as what we would ship to customers.

Reply:
> If, however, this PR is simply to prove out teh API you want ot make public, then this is ok only as something not to be checked in.  Effectively, the PR would be: here's the API shape we think is right, and here's a feature that shows that it works.

Tagging @jinujoseph. The idea is to implement an IDE feature (and few more to follow) based on a possible compiler API and then eventually decide the shape of the eventual compiler API. For example, compiler's current CFG walker does not support backward analysis, and it is very likely we need to support it for future dataflow analyses. Developing these additional IDE features may take us few more cycles/previews and it does not make sense to hold back the useful IDE features in a separate feature branch, while also burdening the feature branch with continuous merge resolutions. Using shared files is an interim step to workaround using a not-yet-baked API instead of maintaining two separate clones in compilers and workspaces layer.

---
In reply to: [229796338](https://github.com/dotnet/roslyn/pull/30777#discussion_r229796338) [](ancestors = 229796338)

Reply:
I will continually push back against shipping solutions to customers that are built this way.  This will simply be used as an escape hatch that eliminates the need for us to actually move the APIs to the public point.

We can use this intenrally.  But until we're certain that these APIs are good enough and ready enough for customers, we shouldn't be shipping our own stuff on top of it.

> and it does not make sense to hold back the useful IDE features in a separate feature branch,

Sure it does.  IMO, we should not be shipping IDE features early if we don't have the APIs ready for them.  This has come up several times in other PRs where we've taken shortcuts to ship things to customers, building up debt and making the codebase more fragile (for example: https://github.com/dotnet/roslyn/issues/29345)

To me, this all feels like ways to shortcut required costs, ship features that people will claim are done, and then never get around to fixing the root problems.  It will almost always be the case that 6-12 months from now there will be little to no management motivation to invest into this debt.  I've heard directly things like "what's the customer ask for something like that?", whereas it's really attractive to people to ship shiny stuff now.  

We've made it this far in roslyn so far without having to take these sorts of shortcuts.  I dont' see why we'd change things now and do this sort of thing, except to subvert a system that was working and had extremely positive externalities that we're now giving up. :-/

Reply:
> IMO, we should not be shipping IDE features early if we don't have the APIs ready for them. 

One **very** important thing that is different about the APIs we are talking about here - these APIs are technically and semantically based in an "Analysis" layer, which is not really part of core compiler, and is completely based on existing public APIs exposed in compiler layer. So technically, we can just have this code in the IDE layer as a dataflow walker based on existing public APIs, but I have taken the current approach just to share code with existing walker in compiler layer. If the team believes it is better to keep these dataflow walkers separate in compiler and IDE layer, I can move the code completely to IDE layer and revert compiler changes, until we have public APIs for dataflow analysis.

Also note that we _have_ taken similar approaches in past during IOperation development. We moved IDE analyzers to be based on **internal** IOperation APIs, and used reflection to invoke them. Once the API was released to public without feature flag, we removed the workaround and moved the IDE analyzers to call into public APIs. This helped improve both the IOperation API and IDE analyzers, and at the same time pushed early adoption of the API. I believe such an approach works very nice for developing/baking APIs that fall in the "analysis/analyzers" category, which in reality sit on top of public compiler APIs, but are still exposed from the Compilers code base.

---
In reply to: [229812940](https://github.com/dotnet/roslyn/pull/30777#discussion_r229812940) [](ancestors = 229812940)

Reply:
> is this an internal file that take compiler logic and tries to embed it directly into the workspace

I believe your confusion is coming from thinking this is some internal compiler logic. The type being shared here is an independent ControlFlowGraph walker utility that allows tracking custom data for each basic block, it does not belong to or is owned by either compiler or workspaces layer as such, but is being used by both in slightly different contexts.

---
In reply to: [229819772](https://github.com/dotnet/roslyn/pull/30777#discussion_r229819772) [](ancestors = 229819772,229812940)

Reply:
> it does not belong to or is owned by either compiler or workspaces layer as such, but is being used by both in slightly different contexts.

Who does it belong to?  How would a 3rd party implement this sort of feature?

Reply:
It belongs to the analysis layer OR the analyzers. So technically, if we were to decide that Microsoft.CodeAnalysis APIs does not intend to provide public support for dataflow walker, then each flow analysis based analyzer will have to do so in it's own package. We already have a similar walker in the roslyn-analyzers repo for flow analysis based on compiler's CFG. However, I believe there is general consensus on working towards making it a public API, but I am not sure about the timelines of this work. Meanwhile, we can either use shared files to avoid duplicate code, or have cloned files in compiler and workspaces layer for doing similar analysis.

Reply:
> owever, I believe there is general consensus on working towards making it a public API, but I am not sure about the timelines of this work. Meanwhile, we can either use shared files to avoid duplicate code, or have cloned files in compiler and workspaces layer for doing similar analysis.

I would definitely go with the latter.  It woudl be an accurate demonstration of the debt we're building by not having this be public somewhere.  As we fixed/changed one, we'd have to understand the downstream work and impact of that.  With linked files, it just becomes too easy to get complacent and feel like things are ok because "we shipped the feature, yaay!", and to never get around to addressing the debt.  

> Also note that we have taken similar approaches in past during IOperation development. We moved IDE analyzers to be based on internal IOperation APIs, and used reflection to invoke them. 

I would also be ok with this approach for hte same reason as above.  it would be very clear there was a significant problem here and that we were having to do very oogy stuff as a short term workaround for it.  This would provide *heavy* pressure on us to be able to actually make the feature public.

For example, IIRC, because of the feature flag, wasn't it the case that the *end user* had to opt-into the flag?  Such a situation would definitely act as a forcing function to complete this work.  Otherwise, we're at the mercy of what we've seen in the past, where we have good intentions to do this work, but others then go "nah... these new features are actually more important".  :-/

--

So, TLDR, i'd take either of the above two approaches over the linked file approach here.

Reply:
> I would definitely go with the latter.  It woudl be an accurate demonstration of the debt we're building by not having this be public somewhere.  As we fixed/changed one, we'd have to understand the downstream work and impact of that.  With linked files, it just becomes too easy to get complacent and feel like things are ok because "we shipped the feature, yaay!", and to never get around to addressing the debt.  

Seems like many other folks invoked in CFG/Flow analysis design agree with you here, so I will revert all compiler changes and move the files directly into the IDE layer.

---
In reply to: [229849136](https://github.com/dotnet/roslyn/pull/30777#discussion_r229849136) [](ancestors = 229849136)

Reply:
Ok.  Thanks for letting me know!  I appreciate all the communication and work on these tough issues.  Thanks!